File: packages/go-sdk/biu/sdkproto/v1/mcp.go
package sdkproto

import (
	"encoding/json"
	"fmt"
)

const (
	McpTypeStdio       = "stdio"
	McpTypeSSE         = "sse"
	McpTypeHTTP        = "http"
	McpTypeSDK         = "sdk"
	McpTypeClaudeProxy = "claudeai-proxy"

	McpStatusConnected = "connected"
	McpStatusFailed    = "failed"
	McpStatusNeedsAuth = "needs-auth"
	McpStatusPending   = "pending"
	McpStatusDisabled  = "disabled"
)

// McpServerConfig 是 5 个 server 类型的标记接口。
type McpServerConfig interface {
	isMcpServerConfig()
	McpServerType() string
}

type McpStdioServerConfig struct {
	Type    string            `json:"type"` // "stdio"
	Command string            `json:"command"`
	Args    []string          `json:"args,omitempty"`
	Env     map[string]string `json:"env,omitempty"`
}

func (*McpStdioServerConfig) isMcpServerConfig()    {}
func (*McpStdioServerConfig) McpServerType() string { return McpTypeStdio }

type McpSSEServerConfig struct {
	Type    string            `json:"type"` // "sse"
	URL     string            `json:"url"`
	Headers map[string]string `json:"headers,omitempty"`
}

func (*McpSSEServerConfig) isMcpServerConfig()    {}
func (*McpSSEServerConfig) McpServerType() string { return McpTypeSSE }

type McpHttpServerConfig struct {
	Type    string            `json:"type"` // "http"
	URL     string            `json:"url"`
	Headers map[string]string `json:"headers,omitempty"`
}

func (*McpHttpServerConfig) isMcpServerConfig()    {}
func (*McpHttpServerConfig) McpServerType() string { return McpTypeHTTP }

type McpSdkServerConfig struct {
	Type     string          `json:"type"` // "sdk"
	Name     string          `json:"name"`
	Instance json.RawMessage `json:"instance,omitempty"`
}

func (*McpSdkServerConfig) isMcpServerConfig()    {}
func (*McpSdkServerConfig) McpServerType() string { return McpTypeSDK }

type McpClaudeAIProxyServerConfig struct {
	Type    string            `json:"type"` // "claudeai-proxy"
	URL     string            `json:"url"`
	Headers map[string]string `json:"headers,omitempty"`
}

func (*McpClaudeAIProxyServerConfig) isMcpServerConfig()    {}
func (*McpClaudeAIProxyServerConfig) McpServerType() string { return McpTypeClaudeProxy }

// UnmarshalMcpServerConfig dispatch 5 个变体。stdio 没有显式 type 字段时也按 stdio 处理（协议默认行为）。
func UnmarshalMcpServerConfig(data []byte) (McpServerConfig, error) {
	var head struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal(data, &head); err != nil {
		return nil, fmt.Errorf("mcp_server: peek type: %w", err)
	}
	var v McpServerConfig
	switch head.Type {
	case "", McpTypeStdio:
		v = &McpStdioServerConfig{}
	case McpTypeSSE:
		v = &McpSSEServerConfig{}
	case McpTypeHTTP:
		v = &McpHttpServerConfig{}
	case McpTypeSDK:
		v = &McpSdkServerConfig{}
	case McpTypeClaudeProxy:
		v = &McpClaudeAIProxyServerConfig{}
	default:
		return nil, fmt.Errorf("mcp_server: unknown type %q", head.Type)
	}
	if err := json.Unmarshal(data, v); err != nil {
		return nil, fmt.Errorf("mcp_server: unmarshal %s: %w", head.Type, err)
	}
	return v, nil
}

type McpServerInfo struct {
	Name    string `json:"name,omitempty"`
	Version string `json:"version,omitempty"`
}

type McpServerStatus struct {
	Name       string         `json:"name"`
	Status     string         `json:"status"` // connected | failed | needs-auth | pending | disabled
	ServerInfo *McpServerInfo `json:"serverInfo,omitempty"`
	Error      string         `json:"error,omitempty"`
}

File: apps/client/lib/data/api/sdkproto/v1/mcp.dart
// 5 个 McpServerConfig 类型 + McpServerStatus

import 'package:json_annotation/json_annotation.dart';

part 'mcp.g.dart';

abstract class McpServerConfig {
  String get type;
  Map<String, dynamic> toJson();

  static McpServerConfig fromJson(Map<String, dynamic> json) {
    final t = json['type'] as String? ?? 'stdio';
    switch (t) {
      case 'stdio':
      case '':
        return McpStdioServerConfig.fromJson(json);
      case 'sse':
        return McpSSEServerConfig.fromJson(json);
      case 'http':
        return McpHttpServerConfig.fromJson(json);
      case 'sdk':
        return McpSdkServerConfig.fromJson(json);
      case 'claudeai-proxy':
        return McpClaudeAIProxyServerConfig.fromJson(json);
      default:
        throw ArgumentError('unknown McpServerConfig type: $t');
    }
  }
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class McpStdioServerConfig implements McpServerConfig {
  @override
  final String type;
  final String command;
  final List<String>? args;
  final Map<String, String>? env;

  McpStdioServerConfig({
    this.type = 'stdio',
    required this.command,
    this.args,
    this.env,
  });

  factory McpStdioServerConfig.fromJson(Map<String, dynamic> json) =>
      _$McpStdioServerConfigFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$McpStdioServerConfigToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class McpSSEServerConfig implements McpServerConfig {
  @override
  final String type;
  final String url;
  final Map<String, String>? headers;

  McpSSEServerConfig({this.type = 'sse', required this.url, this.headers});
  factory McpSSEServerConfig.fromJson(Map<String, dynamic> json) =>
      _$McpSSEServerConfigFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$McpSSEServerConfigToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class McpHttpServerConfig implements McpServerConfig {
  @override
  final String type;
  final String url;
  final Map<String, String>? headers;

  McpHttpServerConfig({this.type = 'http', required this.url, this.headers});
  factory McpHttpServerConfig.fromJson(Map<String, dynamic> json) =>
      _$McpHttpServerConfigFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$McpHttpServerConfigToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class McpSdkServerConfig implements McpServerConfig {
  @override
  final String type;
  final String name;
  final dynamic instance;

  McpSdkServerConfig({this.type = 'sdk', required this.name, this.instance});
  factory McpSdkServerConfig.fromJson(Map<String, dynamic> json) =>
      _$McpSdkServerConfigFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$McpSdkServerConfigToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class McpClaudeAIProxyServerConfig implements McpServerConfig {
  @override
  final String type;
  final String url;
  final Map<String, String>? headers;

  McpClaudeAIProxyServerConfig({
    this.type = 'claudeai-proxy',
    required this.url,
    this.headers,
  });
  factory McpClaudeAIProxyServerConfig.fromJson(Map<String, dynamic> json) =>
      _$McpClaudeAIProxyServerConfigFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$McpClaudeAIProxyServerConfigToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class McpServerInfo {
  final String? name;
  final String? version;

  McpServerInfo({this.name, this.version});

  factory McpServerInfo.fromJson(Map<String, dynamic> json) =>
      _$McpServerInfoFromJson(json);
  Map<String, dynamic> toJson() => _$McpServerInfoToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class McpServerStatus {
  final String name;
  final String status;
  final McpServerInfo? serverInfo;
  final String? error;

  McpServerStatus({
    required this.name,
    required this.status,
    this.serverInfo,
    this.error,
  });

  factory McpServerStatus.fromJson(Map<String, dynamic> json) =>
      _$McpServerStatusFromJson(json);
  Map<String, dynamic> toJson() => _$McpServerStatusToJson(this);
}
